rbx_reflection_database/src/lib.rs

```diff
@@ -1,15 +1,147 @@
+        Ok(None)
+    }
+});
+
+/// Returns a populated [`ReflectionDatabase`]. This will attempt to load one locally and
+/// if one can't be found, it will return one that is bundled with this crate.
+///
+/// ## Errors
+///
+/// Errors if a locally stored [`ReflectionDatabase`] could not be read
+/// or is invalid MessagePack.
+pub fn get() -> Result<&'static ReflectionDatabase<'static>, Error> {
+    Ok(get_local()?.unwrap_or(&BUNDLED_DATABASE))
+}
+
+/// Returns a reflection database from the file system, if one can be found.
+/// This is loaded from a location set by the `RBX_DATABASE` environment
+/// variable if it's set. Otherwise, the default location is checked.
+///
+/// The default location varies depending upon OS:
+///
+/// | OS      | Location                                                            |
+/// |:--------|:--------------------------------------------------------------------|
+/// | Windows | `%localappdata%/.rbxreflection/database.msgpack`                    |
+/// | MacOS   | `$HOME/Library/Application Support/.rbxreflection/database.msgpack` |
+/// | Linux   | `$HOME/.rbxreflection/database.msgpack`                             |
+///
+/// The file at the above location (or the one pointed to by `RBX_DATABASE`)
+/// must be valid MessagePack.
+///
+/// ## Errors
+///
+/// Errors if the file specified by `RBX_DATABASE` or in the default location
+/// exists but is invalid MessagePack.
+pub fn get_local() -> ResultOption<&'static ReflectionDatabase<'static>> {
+    let local_database: &ResultOption<ReflectionDatabase<'static>> = &LOCAL_DATABASE;
+    match local_database {
+        Ok(opt) => Ok(opt.as_ref()),
+        // This clone could be avoided because these references are static,
+        // but it'd involve some indirection and these errors are rare anyway.
+        Err(e) => Err(e.clone()),
+    }
 }
 
-pub fn get() -> &'static ReflectionDatabase<'static> {
-    &DATABASE
+/// Returns the locally bundled [`ReflectionDatabase`]. This database may or may
+/// not be up to date, but it will always exist.
+pub fn get_bundled() -> &'static ReflectionDatabase<'static> {
+    &BUNDLED_DATABASE
+}
+
+/// Fetches the location a [`ReflectionDatabase`] is expected to be loaded from.
+/// This may return [`None`] if the local data directory cannot be found.
+pub fn get_local_location() -> Option<PathBuf> {
+    if let Ok(location) = env::var(OVERRIDE_PATH_VAR) {
+        log::debug!("Using environment variable {OVERRIDE_PATH_VAR} to fetch reflection database");
+        Some(PathBuf::from(location))
+    } else {
+        get_local_location_no_var()
+    }
+}
+
+/// Returns the default local location for the reflection database, without
+/// considering the env variable.
+fn get_local_location_no_var() -> Option<PathBuf> {
+    // Due to concerns about the local data directory existing
+    // on Linux, we use the home directory instead.
+    #[cfg(target_os = "linux")]
+    let mut home = dirs::home_dir()?;
```

Comment:
Can we fall back to the home directory only if the local data directory is None?

Reply:
The concern is that it will often be `Some` but it won't actually exist. This happens in WSL, as an example.